servers/zts/src/main/java/com/yahoo/athenz/zts/ZTSImpl.java

```diff
@@ -834,30 +844,114 @@ ServiceIdentity generateZTSServiceIdentity(com.yahoo.athenz.zms.ServiceIdentity
+
         return ztsService;
     }
-    
+
     String generateServiceIdentityName(String domain, String service) {
         return domain + "." + service;
     }
-    
+
     ServiceIdentity lookupServiceIdentity(DomainData domainData, String serviceName) {
-        
+
         List<com.yahoo.athenz.zms.ServiceIdentity> services = domainData.getServices();
         if (services == null) {
             return null;
         }
-        
+
         for (com.yahoo.athenz.zms.ServiceIdentity service : services) {
             if (service.getName().equalsIgnoreCase(serviceName)) {
                 return generateZTSServiceIdentity(service);
             }
         }
-        
+
         return null;
     }
-    
+
+    protected AthenzJWKConfig getAthenzJWKConfig(ResourceContext ctx) {
+        validateAthenzJWKConfigIsUpdated(ctx);
+        return athenzJWKConfig;
+    }
+
+    protected void validateAthenzJWKConfigIsUpdated(ResourceContext ctx) {
+        boolean rfc = false; // todo: configuration? something else?
+        if ((lastAthenzJWKConfUpdateTime == 0) || (lastAthenzJWKConfUpdateTime + millisBetweenAthenzJWKConfUpdates < System.currentTimeMillis())) {
+            synchronized (pubKeysUpdateMutex) {
+                if (lastAthenzJWKConfUpdateTime == 0 || (lastAthenzJWKConfUpdateTime + millisBetweenAthenzJWKConfUpdates < System.currentTimeMillis())) {
+                    List<PublicKeyEntry> ztsPublicKeys = getServiceIdentity(ctx, ATHENZ_SYS_DOMAIN, ZTS_SERVICE).getPublicKeys();
+                    List<PublicKeyEntry> zmsPublicKeys = getServiceIdentity(ctx, ATHENZ_SYS_DOMAIN, ZMS_SERVICE).getPublicKeys();
+
+                    final List<JWK> ztsJWKList = getJWKList(ztsPublicKeys, rfc);
+                    final List<JWK> zmsJWKList = getJWKList(zmsPublicKeys, rfc);
+
+                    if (ztsJWKList == null || zmsJWKList == null) {
+                        LOGGER.error("ZMS or ZTS JWK List is null, cannot build Athenz JWK config file. ZMS: {}, ZTS: {}", zmsJWKList, ztsJWKList);
+                        return;
+                    }
+                    if (athenzJWKConfig == null) {
+                        athenzJWKConfig = new AthenzJWKConfig();
+                    }
+
+                    AthenzJWKConfig newJWKConf = new AthenzJWKConfig()
+                            .setZtsJWK(new JWKList().setKeys(ztsJWKList))
+                            .setZmsJWK(new JWKList().setKeys(zmsJWKList));
+
+                    if (athenzJWKConfChanged(newJWKConf)) {
+                        long now = System.currentTimeMillis();
+                        newJWKConf.setModified(Timestamp.fromMillis(now));
+                        athenzJWKConfig = newJWKConf;
+                        lastAthenzJWKConfUpdateTime = now;
+                    }
+                }
+            }
+        }
+    }
+
+    protected boolean athenzJWKConfChanged(AthenzJWKConfig newJWKConf) {
+        //TODO: do we have a better way to know these kind of changes?
```

Comment:
the timestamp we send back to the client should be the latest update time of the zms and zts services. so we take a look at the last modified timestamp for those services and which one is older is set as the value instead of the current value. Similarly we should use the same logic to find out if any value was changed or not - just take a look at the last modified timestamp of the zms/zts services.